src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py

```diff
@@ -0,0 +1,319 @@
+            files_found = set.intersection(target_files, set(files))
+            for filename in files_found:
+                semgrep_ruleset_file = os.path.join(root, filename)
+
+                try:
+                    with open(semgrep_ruleset_file, encoding="utf-8") as file:
+                        semgrep_ruleset: dict[str, list] = yaml.safe_load(file.read())
+                except yaml.YAMLError as yaml_error:
+                    error_msg = f"Unable to open semgrep rule file {semgrep_ruleset_file}: {yaml_error}."
+                    logger.debug(error_msg)
+                    raise ConfigurationError(error_msg) from yaml_error
+
+                # should be a top-level key "rules", and then a list of rules (dictionaries) with "id" entries
+                try:
+                    for semgrep_rule in semgrep_ruleset["rules"]:
+                        rule_ids.add(semgrep_rule["id"])
+                except (KeyError, TypeError) as format_error:
+                    error_msg = f"Invalid semgrep rule format for {semgrep_ruleset_file}: {format_error}."
+                    logger.debug(error_msg)
+                    raise ConfigurationError(error_msg) from format_error
+
+            target_files = target_files - files_found
+
+        if target_files:
+            error_msg = f"The following semgrep files were not found in {path}: {target_files}"
+            logger.debug(error_msg)
+            raise ConfigurationError(error_msg)
+
+        return rule_ids
+
+    def analyze(self, pypi_package_json: PyPIPackageJsonAsset) -> tuple[HeuristicResult, dict[str, JsonType]]:
+        """Analyze the source code of the package for malicious patterns.
+
+        This is the first phase of the source code analyzer.
+
+        Parameters
+        ----------
+        pypi_package_json: PyPIPackageJsonAsset
+            The PyPI package JSON asset object.
+
+        Returns
+        -------
+        tuple[HeuristicResult, dict[str, JsonType]]
+            Containing the analysis results and relevant patterns identified.
+
+        Raises
+        ------
+        HeuristicAnalyzerValueError
+            if there is no source code available.
+        """
+        analysis_result: dict = {}
+        disabled_results: dict = (
+            {}
+        )  # since we have to run them anyway, return disabled rule findings for debug information
+        # only run semgrep open-source features, and disable 'nosemgrep' ignoring so this does not bypass our scan
+        semgrep_commands: list[str] = ["semgrep", "scan", "--oss-only", "--disable-nosem"]
+        result: HeuristicResult = HeuristicResult.PASS
+
+        source_code_path = pypi_package_json.package_sourcecode_path
+        if not source_code_path:
+            error_msg = "Unable to retrieve PyPI package source code path"
+            logger.debug(error_msg)
+            raise HeuristicAnalyzerValueError(error_msg)
+
+        semgrep_commands.extend(["--config", self.default_rule_path])
```

Comment:
Just to double check, `semgrep` allows you specify `--config` multiple times, and it will merge them all into a big "configuration"?